ultralytics/engine/exporter.py

```diff
@@ -237,7 +237,7 @@ def __call__(self, model=None):
         y = None
         for _ in range(2):
             y = model(im)  # dry runs
-        if self.args.half and onnx and self.device.type != "cpu":
+        if self.args.half and (onnx or engine) and self.device.type != "cpu":
```

Comment:
@Burhan-Q failure might be because of this. @ZouJiu1 was saying that without this the inputs and outputs remains FP32 (but the inner layers are FP16 I think).

Can you try with and without this line?

Reply:
@glenn-jocher yes, it appears that commenting this line and using `if self.args.half and onnx and self.device.type != "cpu":` instead works for both TensorRT >= 10.0 and TensorRT < 10. I may have misunderstood, but I think that the inputs (at least) for TensorRT are supposed to be FP32 and that the conversion happens internally, but I can't find where I read this so it might not be correct.

Reply:
hi, @Burhan-Q I think using "(onnx or engine)" is necessary, the conversion with 10.0.0b6 is not internal. Because when use the converted engine tensorrt file,  the codes use input dtype to judge FP32 or FP16, not the arg.half flag, in line [autobackend.py#L244](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/autobackend.py#L244), [autobackend.py#L249](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/autobackend.py#L249) and [autobackend.py#L240](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/autobackend.py#L240). If I use `if self.args.half and onnx and self.device.type != "cpu":` to convert engine file with 10.0.0b6, the converted engine input shows FP32 not FP16 when I debug into [autobackend.py#L244](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/autobackend.py#L244).

So, I think `if self.args.half and (onnx or engine) and self.device.type != "cpu":` is necessary. the conversion is not internal.

Reply:
@ZouJiu1 this single change causes the [error documented here](https://github.com/ultralytics/ultralytics/pull/9516#pullrequestreview-1985271931). I get what you're saying, but it doesn't work when this line is used. We can always update to get the functionality you're looking for after we have solid compatibility with new and older versions of TensorRT